cpp/src/arrow/compute/kernels/scalar_hash.cc

```diff
@@ -0,0 +1,204 @@
+    std::vector<std::shared_ptr<ArrayData>> children;
+
+    auto type_id = array.type->id();
+    if (type_id == Type::STRUCT) {
+      columns.reserve(array.child_data.size());
+      for (size_t i = 0; i < array.child_data.size(); i++) {
+        auto child = array.child_data[i];
+        if (is_nested(child.type->id())) {
+          ARROW_ASSIGN_OR_RAISE(auto child_hashes,
+                                HashChild(array, child, hash_ctx, memory_pool));
+          children.push_back(child_hashes);
+          ARROW_ASSIGN_OR_RAISE(columns[i], ToColumnArray(*child_hashes, hash_ctx));
+        } else {
+          ARROW_ASSIGN_OR_RAISE(columns[i], ToColumnArray(child, hash_ctx));
+        }
+      }
+    } else if (is_list_like(type_id)) {
+      auto values = array.child_data[0];
+      ARROW_ASSIGN_OR_RAISE(auto value_hashes,
+                            HashChild(array, values, hash_ctx, memory_pool));
+      children.push_back(value_hashes);
+      ARROW_ASSIGN_OR_RAISE(
+          columns[0], ToColumnArray(array, hash_ctx, value_hashes->buffers[1]->data()));
+    } else {
+      ARROW_ASSIGN_OR_RAISE(columns[0], ToColumnArray(array, hash_ctx));
+    }
+    Hashing64::HashMultiColumn(columns, hash_ctx, out);
+    return Status::OK();
+  }
+
+  static Status Exec(KernelContext* ctx, const ExecSpan& input_arg, ExecResult* out) {
+    if (input_arg.num_values() != 1 || !input_arg[0].is_array()) {
+      return Status::Invalid("FastHash currently supports a single array input");
+    }
+    ArraySpan hash_input = input_arg[0].array;
+
+    auto exec_ctx = default_exec_context();
+    if (ctx && ctx->exec_context()) {
+      exec_ctx = ctx->exec_context();
+    }
+
+    // Initialize stack-based memory allocator used by Hashing32 and Hashing64
+    util::TempVectorStack stack_memallocator;
+    ARROW_RETURN_NOT_OK(
+        stack_memallocator.Init(exec_ctx->memory_pool(),
+                                3 * sizeof(int32_t) * util::MiniBatch::kMiniBatchLength));
+
+    // Prepare context used by Hashing32 and Hashing64
+    LightContext hash_ctx;
+    hash_ctx.hardware_flags = exec_ctx->cpu_info()->hardware_flags();
+    hash_ctx.stack = &stack_memallocator;
+
+    // Call the hashing function, overloaded based on OutputCType
+    ArraySpan* result_span = out->array_span_mutable();
+    uint64_t* result_ptr = result_span->GetValues<uint64_t>(1);
+    ARROW_RETURN_NOT_OK(
+        HashArray(hash_input, &hash_ctx, exec_ctx->memory_pool(), result_ptr));
+
+    return Status::OK();
+  }
+};
+
+// ------------------------------
+// Function construction and kernel registration
+std::shared_ptr<ScalarFunction> RegisterKernelsFastHash64() {
+  // Create function instance
+  auto fn_hash_64 =
+      std::make_shared<ScalarFunction>("hash_64", Arity::Unary(), hash_64_doc);
+
+  // Add 64-bit hash kernel
+  ScalarKernel kernel({InputType()}, OutputType(uint64()), FastHashScalar::Exec);
+  kernel.null_handling = NullHandling::OUTPUT_NOT_NULL;
```

Comment:
Is this correct? The function doc says this function emits null for null input so this should be `NullHandling::INTERSECTION` (defaulted)?

Reply:
The docs are coming from a previous PR and I haven't updated them yet.

Reply:
Updated the docs.